q2_mystery_stew/plugin_setup.py

```diff
@@ -18,3 +32,229 @@
+                     (qiime2.NumericMetadataColumn(mdc_num_val),))
+}
+
+num_functions = 0
+signatures = generate_signatures()
+for sig in signatures:
+    for param_set in product(all_params.values(), repeat=sig.num_inputs):
+        for params in zip(param_set):
+            action_name = f'func_{num_functions}'
+
+            param_dict = {}
+            for i, param in enumerate(params):
+                param = all_params[param.base_name]
+                param_dict.update({param.base_name + f'_{i}': param})
+
+            # This dictionary allows for easy registration of Qiime parameters
+            param_name_to_type_dict = \
+                {name: value.type for name, value in param_dict.items()}
+
+            rewrite_function_signature(sig.template,
+                                       {},
+                                       param_name_to_type_dict,
+                                       sig.num_outputs,
+                                       action_name)
+
+            outputs = []
+            for i in range(sig.num_outputs):
+                outputs.append((f'output_{i + 1}', EchoOutput))
+
+            chosen_param_values = {}
+            for index, (name, value) in enumerate(param_dict.items()):
+                if value.type == List[Int] or value.type == List[Float]:
+                    param_val = []
+                    for val in range(sig.num_outputs):
+                        domain_size = len(value.domain)
+                        selected_param = \
+                            value.domain[domain_size % sig.num_outputs]
+
+                        selected_size = len(selected_param.domain)
+                        param_val.append(
+                            selected_param.domain[selected_size %
+                                                  sig.num_outputs])
+
+                    chosen_param_values.update({name: param_val})
+                elif value.type == Set[Int] or value.type == Set[Float]:
+                    param_val = set()
+                    for val in range(sig.num_outputs):
+                        domain_size = len(value.domain)
+                        selected_param = \
+                            value.domain[domain_size % sig.num_outputs]
+
+                        selected_size = len(selected_param.domain)
+                        param_val.add(
+                            selected_param.domain[selected_size %
+                                                  sig.num_outputs])
+                    param_val = tuple(param_val)
+
+                    chosen_param_values.update({name: param_val})
+                else:
+                    domain_size = len(value.domain)
+                    chosen_param_values.update(
+                        {name: value.domain[sig.num_outputs % domain_size]})
+
+            usage_example = \
+                {action_name:
+                 UsageInstantiator(chosen_param_values, outputs, action_name)}
```

Comment:
I think we want one of these for each element in a type's domain. This gets wild with multiple inputs, so perhaps fix all other inputs at a single value while iterating through the domain of a given input, that way the number of usages matches the number of values if they were all independent.

Reply:
And do this for each combination of arguments we're generating? This caused me to have an idea. I'm not sure if this is what you meant or not, but what if we only had one function per number of outputs that took every parameter. Then we do what you said above. We create usage examples with every value in an individual argument's domain while holding all others in place. We do this for all arguments.

Additionally, under the current implementation we will end up with a usage example somewhere containing each value in a type's domain provided the domain is three or less elements (which they all are at the moment). It wouldn't be too hard to ensure a usage example if we were to end up with a parameter with a larger domain.